ESQL: Fix `mv_expand` inconsistent column order

kanoshiou opened the pull request:
The new attribute generated by `mv_expand` should remain in the original field positions. However, when `ProjectAwayColumns` is executed top-down and encounters an `MvExpandExec` with the child plan of `ExchangeExec`, the inserted `Project` may not respect that order.

Closes #129000

kanoshiou committed:
Fix mv_expand inconsistent column order

kanoshiou committed:
Update docs/changelog/129745.yaml

kanoshiou committed:
Merge branch 'refs/heads/main' into fix-inconsistent-column-order

# Conflicts:
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlCapabilities.java

kanoshiou committed:
Merge branch 'main' into fix-inconsistent-column-order

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/ProjectAwayColumns.java:
```diff
@@ -49,6 +50,17 @@ public PhysicalPlan apply(PhysicalPlan plan) {
             }
 
+            // for mv_expand, the target attribute should be retained in its original position
+            if (currentPlanNode instanceof MvExpandExec mvExpand) {
```
I don't think we should special case for MvExpand; I believe there is a general bug that can re-occur whenever a plan node adds a new attribute somewhere else than on the right hand side, resp. when the new attributes don't come after all previously existing attributes in the `.output()` list.

In fact, it's a nice coincidence that this rule worked for so long without running into issues, and it probably only worked because most of our plans do append new attributes to the right.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/ProjectAwayColumns.java:
IMO, the problem is in lines 43 (the invariant claimed in the comment!) and 93 (where the new `Project` is constructed).

The invariant lies in our faces: It says it's okay to add a `Project` with the attributes from the `requiredAttrBuilder` after the current plan node, but this only holds up to permutation of the attributes. When traversing the plan, we clearly treat the `requiredAttrBuilder` as an unordered set when we add the current node's references in line 104.

I believe the proper fix is as follows: When we construct the `Project` that we place at the top of the fragment, we should respect the order of the attributes. Since all attributes in `requiredAttrBuilder` have to be in the `.output()` of the top node of the fragment, we can simply use the `.output()` list to determine the order the attributes have to be in; then the `Project` can be constructed correctly, i.e. containing only the used attributes while still preserving the relative order of the attributes.

alex-spies reviewed (changes requested):
Super nice find, thank you @kanoshiou  !

I think the fix would work, but I'd like to attempt a fix that will also work for other plan nodes that we add in the future and might work similarly to `MV_EXPAND`; see my suggestion below.

alex-spies commented:
Thanks again for your contribution @kanoshiou !

Do you mind if we push to your PR in order to try and augment the fix as described [here](https://github.com/elastic/elasticsearch/pull/129745#discussion_r2207954319)?

kanoshiou committed:
Merge branch 'refs/heads/main' into fix-inconsistent-column-order

kanoshiou committed:
Respect the order of the attributes

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/PhysicalPlanOptimizerTests.java:
```diff
@@ -5698,9 +5749,9 @@ public void testPushTopNKeywordToSource() {
         var exchange = asRemoteExchange(topN.child());
```
nit: for the tests in with changes in this file, it'd be nice to update the javadoc that contains the expected optimized plan. A simple way to obtain this is to debug the test in IntelliJ and look at `optimized` via the debugger.

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/PhysicalPlanOptimizerTests.java:
```diff
@@ -5698,9 +5749,9 @@ public void testPushTopNKeywordToSource() {
 
         project = as(exchange.child(), ProjectExec.class);
-        assertThat(names(project.projections()), contains("abbrev", "name", "location", "country", "city"));
+        assertThat(names(project.projections()), contains("abbrev", "city", "country", "location", "name"));
```
Nice. The only reason these queries produced correct output was because there was a `Project` at the top that made the attribute order here irrelevant; but actually, we had changed it locally, and this shows that `ProjectAwayColumns` was never fully correct to begin with.

alex-spies reviewed (commented):
Great, @kanoshiou , thanks for the iteration! This is very close. I only have a final set of remarks; once the remarks in `ProjectAwayColumns.java` are addressed, this LGTM.

kanoshiou commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/ProjectAwayColumns.java:
```diff
@@ -76,7 +76,15 @@ public PhysicalPlan apply(PhysicalPlan plan) {
+                        for (Attribute attribute : logicalFragment.outputSet()) {
+                            if (requiredAttrBuilder.contains(attribute)) {
+                                output.add(attribute);
+                                requiredAttrBuilder.remove(attribute);
+                            }
+                        }
+                        output.addAll(requiredAttrBuilder.build());
```
> all the required attributes *have* to be in the fragment's `outputSet`

I gave it some thought, and it does make sense.

>assert that the size of `output` and the size of `requiredAttrBuilder` are the same

I have a question: my understanding is that `requiredAttrBuilder` should be a subset of `output`, meaning its size should be less than or equal to that of `output`. Why, then, are they the same size here? If their sizes are the same, then why bother maintaining `requiredAttrBuilder`? Wouldn't it be ok to just use `output` for execution?

alex-spies replied:
Well, I was wrong. There's a bug in the planning of remote enrich that can make it so that the output of the fragment doesn't contain an attribute, but it's actually physically there and the projection needs to include it, or a downstream `TopN` won't know what attribute to sort on. This is what happened in https://github.com/elastic/elasticsearch/issues/118531.

@kanoshiou , I'll re-instate your initial approach (but still using `output` rather than `outputSet`) and I'll add a comment to the problem.

alex-spies committed:
Use simpler indices in csv tests

Indices used in enrich policies, as lookup indices, or multiple indices
in the FROM pattern are all treated differently by different test
suites.

alex-spies commented:
@elasticmachine test this please

kanoshiou committed:
Merge branch 'refs/heads/main' into fix-inconsistent-column-order

kanoshiou committed:
Update

alex-spies reviewed (approved):
This is great, thanks @kanoshiou !

alex-spies committed:
Merge branch 'main' into fix-inconsistent-column-order

alex-spies committed:
Revert to kanoshiou's initial approach

Keep all the attributes from requiredAttrBuilder in the projection, even
when they're not in the fragment's output due to a bug.

alex-spies committed:
Merge branch 'main' into fix-inconsistent-column-order

kanoshiou commented:
Oh, I forgot to remove [this line](https://github.com/kanoshiou/elasticsearch/blob/e4ff3f8bae28b78cfd10770fe076309296d8042c/x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/generative/GenerativeRestTest.java#L56). It seems the CI test has not started yet. I'll update the branch shortly.

kanoshiou committed:
Remove 129000

alex-spies committed:
Merge branch 'main' into fix-inconsistent-column-order

alex-spies commented:
@elasticmachine test this please

alex-spies commented:
Here we go! Thanks again @kanoshiou !

alex-spies commented:
## 💔 Some backports could not be created

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|9.0|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/131428">](https://github.com/elastic/elasticsearch/pull/131428)|
|✅|8.19|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/131447">](https://github.com/elastic/elasticsearch/pull/131447)|
|❌|8.18|Could not create pull request: fetch failed|

### Manual backport
To create the backport manually run:
```
backport --pr 129745
```

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 9.4.3 -->
